resources/aws_ec2.go

```diff
@@ -503,71 +280,67 @@ func (obj *AwsEc2Res) longpollWatch() error {
-// detect ec2 instance state changes.
-func (obj *AwsEc2Res) snsWatch() error {
-	send := false
-	var exit *error
-	defer obj.wg.Wait()
-	defer close(obj.closeChan)
-	// set up the sns endpoint
-	snsServer := obj.snsServer()
-	// shutdown the sns endpoint when we're done
-	defer func() {
-		ctx, cancel := context.WithTimeout(context.TODO(), SnsServerShutdownTimeout*time.Second)
-		defer cancel()
-		if err := snsServer.Shutdown(ctx); err != nil {
-			if err != context.Canceled {
-				log.Printf("%s: error stopping sns endpoint: %s", obj, err)
-				return
-			}
-			log.Printf("%s: sns server shutdown cancelled", obj)
-		}
-	}()
-	obj.wg.Add(1)
-	// start the endpoint
-	go func() {
-		defer obj.wg.Done()
-		defer close(obj.awsChan)
-		if err := snsServer.ListenAndServe(); err != nil {
-			// when we shut down
-			if err == http.ErrServerClosed {
-				log.Printf("%s: Stopped SNS Endpoint", obj)
-				return
-			}
-			// any other error
-			select {
-			case obj.awsChan <- &chanStruct{
-				err: errwrap.Wrapf(err, "sns server error"),
-			}:
-			case <-obj.closeChan:
-			}
+// watchWaiter waits for an instance to change state, and returns the state.
+func watchWaiter(ctx context.Context, instance *ec2.Instance, c *ec2.EC2) (string, error) {
+	var err error
+	var name string
+
+	// these cases are not permitted.
+	if instance == nil {
+		return "", fmt.Errorf("nil instance")
+	}
+	if aws.StringValue(instance.State.Name) == "" {
+		return "", fmt.Errorf("nil or empty state")
+	}
+
+	// get the instance name.
+	for _, tag := range instance.Tags {
+		if aws.StringValue(tag.Key) == "Name" {
+			name = aws.StringValue(tag.Value)
 		}
-	}()
-	log.Printf("%s: Started SNS Endpoint", obj)
-	// process events
-	for {
-		select {
-		case event := <-obj.Events():
-			if exit, send = obj.ReadEvent(event); exit != nil {
-				return *exit
-			}
-		case msg, ok := <-obj.awsChan:
-			if !ok {
-				return *exit
-			}
-			if err := msg.err; err != nil {
-				return err
-			}
-			log.Printf("%s: State: %v", obj, msg.event)
-			obj.StateOK(false)
-			send = true
+	}
+	// error if we didn't find one.
+	if name == "" {
+		return "", fmt.Errorf("name not found")
+	}
+
+	// instance.InstanceId might be nil in case of terminated instances, which
+	// the api will treat as if it isn't included.
+	waitInput := &ec2.DescribeInstancesInput{
+		InstanceIds: []*string{instance.InstanceId},
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("tag:Name"),
+				Values: []*string{aws.String(name)},
+			},
+		},
+	}
+	// call on the appropriate waiter based on current state.
+	switch aws.StringValue(instance.State.Name) {
+	case ec2.InstanceStateNameRunning, ec2.InstanceStateNameStopping:
+		err = c.WaitUntilInstanceStoppedWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameStopped, ec2.InstanceStateNamePending:
+		err = c.WaitUntilInstanceRunningWithContext(ctx, waitInput)
+	case ec2.InstanceStateNameTerminated:
+		err = c.WaitUntilInstanceExistsWithContext(ctx, waitInput)
```

Comment:
There's a race where this returns some milliseconds before the instance is available (according to my describeInstanceByName function,) so we get 4 or 5 extra "terminated" events. I know sleeping for a second isn't a great fix, but it _would_ work in this case. Better ideas?

Reply:
This waiter doesn't work the way I thought it did. Troubleshooting it now.

Reply:
Figured it out, I was waiting for any instance with the right name to exist, including terminated ones. ಠ_ಠ
I think I made a change that stopped it waiting for terminated ones in the original version, which might be why waituntilinstancestopped wasn't behaving how I thought.